modules/common/cronjob/cronjob.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2022 Red Hat
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cronjob
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	batchv1 "k8s.io/api/batch/v1"
+	k8s_errors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/openstack-k8s-operators/lib-common/modules/common/helper"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/util"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// NewCronJob returns an initialized CronJob.
+func NewCronJob(
+	cronjob *batchv1.CronJob,
+	timeout int,
+) *CronJob {
+	return &CronJob{
+		cronjob: cronjob,
+		timeout: time.Duration(timeout) * time.Second,
+	}
+}
+
+// CreateOrPatch - creates or patches a cronjob, reconciles after Xs if object won't exist.
+func (cj *CronJob) CreateOrPatch(
+	ctx context.Context,
+	h *helper.Helper,
+) (ctrl.Result, error) {
+	op, err := controllerutil.CreateOrPatch(ctx, h.GetClient(), cj.cronjob, func() error {
+		err := controllerutil.SetControllerReference(h.GetBeforeObject(), cj.cronjob, h.GetScheme())
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+	if err != nil {
+		if k8s_errors.IsNotFound(err) {
+			h.GetLogger().Info(fmt.Sprintf("CronJob %s not found, reconcile in %s", cj.cronjob.Name, cj.timeout))
+			return ctrl.Result{RequeueAfter: cj.timeout}, nil
+		}
+		return ctrl.Result{}, err
+	}
+	if op != controllerutil.OperationResultNone {
+		h.GetLogger().Info(fmt.Sprintf("CronJob %s - %s", cj.cronjob.Name, op))
+		return ctrl.Result{RequeueAfter: cj.timeout}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// Delete - delete a cronjob.
+func (cj *CronJob) Delete(
+	ctx context.Context,
+	h *helper.Helper,
+) error {
+	err := h.GetClient().Delete(ctx, cj.cronjob)
+	if err != nil && !k8s_errors.IsNotFound(err) {
+		err = fmt.Errorf("Error deleting cronjob %s: %v", cj.cronjob.Name, err)
+		return err
+	}
+
+	return nil
+}
+
+// GetCronJob - get the cronjob object.
+func (cj *CronJob) GetCronJob() batchv1.CronJob {
+	return *cj.cronjob
+}
+
+// SetTimeout defines the duration used for requeueing while waiting for the cronjob
+// to finish.
+func (cj *CronJob) SetTimeout(timeout time.Duration) {
+	cj.timeout = timeout
+}
+
+// GetCronJobWithName func
+func GetCronJobWithName(
+	ctx context.Context,
+	h *helper.Helper,
+	name string,
+	namespace string,
+) (*batchv1.CronJob, error) {
```

Comment:
I'm fine with updating the function accordingly. My main aim is to provide the consistent interface with the other modules.

Reply:
To make sure my understanding is correct, I assume the expectation is to add GetCronJobByName function, which internally use GetCronJobWithName, right ? If that is correct then I'd prefer leaving that now, since we are not really using Get*WithName functions in other operators atm and adding a new interface does not affect subsequent implementation in operators side.

Reply:
I agree with @stuggi here.

Reply:
@kajinamit your understanding is correct. Have GetCronJobByName() like https://github.com/openstack-k8s-operators/lib-common/blob/master/modules/database/funcs.go#L312 and use a getCronJobWithName like https://github.com/openstack-k8s-operators/lib-common/blob/master/modules/database/funcs.go#L268.  I think we should already do this for CronJob and then enhance the others first by adding GetCronJobByName and when all consumers got updated then to the adjustment to lib-common.

Reply:
I updated the PR to include GetCronJobByName

Reply:
One note about https://github.com/openstack-k8s-operators/lib-common/blob/master/modules/database/funcs.go#L312 that the name there is the dbName instead of the name of the k8s resource, i.e. metadata.Name. It think that is a sub-optimal simplification there. So when we follow this pattern please try to idenify CronJob by the resource name.

Reply:
IIUC the current proposed implementation should look up the resource by resource name and namespace name but please correct me if I'm wrong .

Reply:
yes, this code does that. thanks.